### ISSUE TITLE
networkStatus gets stuck on "2" (setVariables)

### ISSUE THREAD

**Issue by SachaG** [apollographql/react-apollo#667]
Something I've noticed recently: when passing new options to my query container, it can get stuck on `networkStatus = 2` even once the request succeeds. Here's a short screencast to demonstrate the issue:

https://cl.ly/kEWd

Note that this only seems to happen the first time these new options are passed. If I go back and forth between both pages, it'll work the second time. This is with apollo-client 1.1.2 and react-apollo 1.1.1.

Other (maybe related) `networkStatus` issues: 

https://github.com/apollographql/apollo-client/issues/1186
https://github.com/apollographql/react-apollo/issues/170

**Comment by grossberg** [comment]
i have the same issue

**Comment by Josiang** [comment]
i also have the exact same issue

**Comment by helfer** [comment]
Thanks @SachaG! Would anyone be able to provide a reproduction using react-apollo-error-template? 🙂

**Comment by hiei189** [comment]
Is there a workaround? like using another version?

**Comment by alikhani** [comment]
I'm having the same issue.. Any progress?

**Comment by helfer** [comment]
@hiei189  @alikhani I think this issue is similar to another one we're trying to debug (where the loading state doesn't change back), but we could still use a reproduction for this, in case it's not the same.

**Comment by andreasgruenh** [comment]
Hello everybody, I encountered this problem as well and I was able to reproduce this in this repo: https://github.com/andreasgruenh/react-apollo-error-template

Take a look at the PersonDisplay component. It receives an ID as prop and should render the corresponding person. To prevent some old state in the PersonDisplay component, I want it to completly unmount and remount it (by giving it a key prop). When I give this prop from the App component, everything works, when I hand this prop from mapProps HOC, it stops working. It also does not happen with react-apollo 1.0.0

**Comment by alikhani** [comment]
I can try to reproduce it later when I have time. I can confirm that the issue manifests itself in react-apollo@1.0.1. So 1.0.0 works fine.

**Comment by helfer** [comment]
@alikhani what's the latest version of apollo-client (not react-apollo) that it works with and the first version where it breaks?

**Comment by alikhani** [comment]
@helfer I'm importing apollo-client from react-apollo. Not sure how the dependencies work.. 
react-apollo@1.0.0 that works for me has the dependency: apollo-client "^1.0.0-rc.9". In my node_modules the version I find is apollo-client 1.2.2.

In react-apollo 1.0.0 where it breaks it says that it should use "^1.0.2" [#L132](https://github.com/apollographql/react-apollo/blob/d3fc77285769f34c2f14960632e7fadb0be0cea2/package.json#L132)

So I think I'm using the latest apollo-client release(1.2.2), due to the "^", on both versions.

**Comment by helfer** [comment]
@alikhani can you try this again after you set the version in your package.json explicitly `react-apollo@^1.2.0`, then `rm -rf node_modules && npm install`. Then `npm list | grep apollo` should tell you which versions you have.

**Comment by jole78** [comment]
I can also confirm that this is an issue. However, I got it working by downgrading react-apollo to 1.0.0

This is my setup (working):
apollo-client@1.2.2
react-apollo@1.0.0


Also using:
fetchPolicy: "network-only"

Upgrading react-apollo makes the network status stuck in "2", thus loading is always true and my UI is showing the "loading spinner".

**Comment by helfer** [comment]
@jole78 interesting! Can you tell us at exactly which version of react-apollo things start breaking? That will help us narrow down which change caused this.

**Comment by hiei189** [comment]
I get this when running `yarn list | grep apollo`

├─ apollo-client@1.0.3
├─ react-apollo@1.0.2
│  ├─ apollo-client@^1.0.2

and it is working ok.

**Comment by jole78** [comment]
@helfer sorry for the late response. Anything higher than 1.0.0

**Comment by rovansteen** [comment]
I can confirm I have this issue too. Had to downgrade to 1.0.* for it to work properly.

**Comment by c10b10** [comment]
I can confirm the issue as well.

**Comment by kolpav** [comment]
This problem is driving me crazy. If there is any way how I can provide some information to help you narrow down the problem let me know.

**Comment by bogdansoare** [comment]
Also have this problem

**Comment by igorlamos** [comment]
Same for me ... thanks for info about downgrade, which also solved the problem for me.

**The following works fine**:

```
├─ apollo-client@1.3.0
├─ react-apollo@1.0.0
│  ├─ apollo-client@^1.0.0-rc.9
```

**The following does not work**:

```
├─ apollo-client@1.3.0
├─ react-apollo@1.0.1
│  ├─ apollo-client@^1.0.2
```

**Comment by jbaxleyiii** [comment]
@helfer yeah recycled queries felt like it added in a lot of complexity / bugs. I wonder it it's worth taking a look at a different approach?

**Comment by helfer** [comment]
@jbaxleyiii At this point I think recycled queries are there just for backwards compatibility to support updating queries that aren't currently being used. For folks that are using only `update` or who are never updating the cache, we could turn them off completely. We can make them opt-out for now, and for the next major version of react-apollo we'll make them opt-in. Do you see any issues with that approach?

**Comment by jbaxleyiii** [comment]
@helfer I like it!

**Comment by andreasgruenh** [comment]
Could you elaborate on the meaning of "queries that aren't beeing used"? Let's say I have 2 routes, on one route is a list of all cars, on the other Route there is a button to add a car. Can I  still update the cars list while beeing on the add car site (list site is not mounted)?

**Comment by SachaG** [comment]
> Can I still update the cars list while being on the add car site (list site is not mounted)?

@helfer will correct me if I'm wrong, but I think the answer is that without recycled queries, you **can** do this if you use `update`, but you **can't** if you're using the (probably soon deprecated if they aren't already) `updateQueries` or a reducer.

**Comment by helfer** [comment]
@andreasgruenh It's basically what @SachaG said. You can't reference the query by name, but you can update the cache using the `update` function. This is because if the query isn't active any more, it will no longer be tracked and thus can't be updated by name.

**Comment by jbaxleyiii** [comment]
good news everyone! This will be fixed in 1.4.0 which should release tomorrow!

**Comment by jbaxleyiii** [comment]
This is now fixed via 1.4.0! You will need to update apollo-client to the same version!

**Comment by igorlamos** [comment]
GREAT! Works for me ... Updated to:
```
├─ react-apollo@1.4.0
│  ├─ apollo-client@^1.4.0
```

and problem solved! **Awesome work** 👍

**Comment by slackday** [comment]
I can confirm this works with `react-apollo: 1.4.2` and `apollo-client: 1.4.0`.

I ran into this issue when refactoring my code from using multiple graphql queries to a single. Thought I had some error in my components and without being able to resolve it myself I started looking for similar issues and found this.

network status got set to 2 and being stuck there when only changing the query variables calculated from props. I see it being related to query re-use which makes sense.

Great job everyone in resolving this! And thanks :)
